OpInfo for nn.functional.interpolate

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64214 Opinfos for avg_pooling
* #61971 nn.functional.linear OpInfo
* **#61956 OpInfo for nn.functional.interpolate**

Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

Differential Revision: [D30013751](https://our.internmc.facebook.com/intern/diff/D30013751)

zou3519 committed:
OpInfo for nn.functional.interpolate

Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfo for nn.functional.interpolate"


Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfo for nn.functional.interpolate"


Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 commented:
@jbschlosser and I discussed a bit offline:
- in the near-to-mid-term future we want to autogenerate nn functional OpInfos from ModuleInfos.
- Until that happens, because functorch needs OpInfo tests in the short-term to test features, I'm manually writing some regular OpInfos for nn functionals and submitting them as changes to PyTorch
- When autogeneration of functional OpInfos from ModuleInfos works, we'll migrate these to ModuleInfos (and any future things functorch needs tests for I'll add as ModuleInfos)

jbschlosser commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6256,6 +6299,77 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            autodiff_nonfusible_nodes=["aten::hardswish"]),
+    OpInfo('nn.functional.interpolate',
+           aten_name="interpolate",
+           variant_test_name='nearest',
```
Is the purpose of adding multiple interpolation-type variants because:
* Different dtypes are supported for each type of interpolation
* Some interpolation types are supported on CUDA and some aren't
* Any other reason?

zou3519 replied:
The main reason is that each variant calls a different backend: https://github.com/pytorch/pytorch/blob/a8f6b5a80a48ed814443c36d40f9eaf5c6aef7f6/torch/nn/functional.py#L3689-L3708 .

This leads to weird behavior that different dtypes are supported for each type of interpolation on different devices

jbschlosser replied:
Gotcha- thanks for the clarification. This is a pretty interesting case for ModuleInfo & functional OpInfo generation - the single module `nn.Upsample` encapsulates all these different ops with different dtype / device support.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2158,6 +2158,49 @@ def sample_inputs_hardswish(self, device, dtype, requires_grad):
+    for align_corners in align_corners_options:
+        for rank in ranks_for_mode[mode]:
+            sample_inputs.extend([
+                SampleInput(make_tensor(shape(D, rank), device=device, dtype=dtype,
```
Not a requirement, but one trick people use to implement these more succinctly is to partial make_tensor with args like device, dtype, requires_grad, low and high to avoid writing them all the time.

zou3519 replied:
Thanks for pointing out the trick!

mruberry reviewed (approved):
A tricky OpInfo to add with the divergence code paths and dtype support! Nice work, @zou3519. It would be nice if the code paths supported the same dtypes.

I made one inline style suggestion that might save you a little time in the future on the sample_inputs. I trust you've designed them carefully to test interesting cases. The OpInfos themselves LGTM!

zou3519 committed:
Update on "OpInfo for nn.functional.interpolate"


Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfo for nn.functional.interpolate"


Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30013751).

zou3519 committed:
Update on "OpInfo for nn.functional.interpolate"


Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

Differential Revision: [D30013751](https://our.internmc.facebook.com/intern/diff/D30013751)

[ghstack-poisoned]

zou3519 committed:
Update on "OpInfo for nn.functional.interpolate"


Each mode goes through a different implementation so they are listed as
different variants.

Test Plan:
- run tests

Differential Revision: [D30013751](https://our.internmc.facebook.com/intern/diff/D30013751)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30013751).
